imageJPG.ml

```diff
@@ -17,71 +17,373 @@
-    let headers = read_header_data ich in
-    close_chunk_reader ich;
-
-    let rec find_SOF ls =
-      match ls with
-      | []        -> raise (Corrupted_image "No SOFn marker...");
-      | (m,d)::bs -> if m >= 0xc0 && m <= 0xcf && m <> 0xc4 && m <> 0xcc
-                     then (m - 0xc0, d)
-                     else find_SOF bs
+    let rec handle_marker ich c =
+      match string_of_marker c with
+      | "EOI" -> raise Not_found
+      | "SOF0" | "SOF1"  | "SOF2"  | "SOF3"  | "SOF5"  | "SOF6"  | "SOF7"
+      | "SOF9" | "SOF10" | "SOF11" | "SOF13" | "SOF14" | "SOF15" ->
+        let sof = parse_sof ich in
+        (sof.width, sof.height)
+      | marker ->
+        (if not (marker_stands_alone marker) then
+           let len = get_bytes ich 2 |> int_of_str2_be in
+           let data = get_bytes ich (len - 2) in
+           ignore data
+        );
+        find_next_marker ich handle_marker
     in
-    let (n, sof) = find_SOF headers in
-    (*
-    Printf.fprintf stderr "SOF%i: " n;
-    show_string_hex sof;
-    *)
-    let height = ((int_of_char sof.[1]) lsl 8) lor (int_of_char sof.[2]) in
-    let width = ((int_of_char sof.[3]) lsl 8) lor (int_of_char sof.[4]) in
-    width, height
+    try
+      find_next_marker ich handle_marker
+    with End_of_file ->
+      raise (Corrupted_image "Reached end of file while looking for SOF marker")
 
   let parsefile fn =
-    raise (Not_yet_implemented "ImageJPG.openfile") (* TODO  *)
+    let make_dqt precision index entries =
+      { precision; index; entries }
+    in
+
+    let read_chunks = ref [] in
+
+    let sof = ref None in
+    let dqts = [| make_dqt 8 0 default_luminance_quantize_table; make_dqt 8 1 default_chrominance_quantize_table |] in
+    let sos = ref None in
+    let image = ref None in
+
+    let rec handle_marker ich c =
+      let curr_ctype = string_of_marker c in
+      if !debug then
+        Printf.eprintf "Found marker %s\n%!" curr_ctype; (
+        match curr_ctype with
+        (* Required markers *)
+        | "SOI" ->
+          only_once ich read_chunks curr_ctype;
+          only_before ich read_chunks "SOS" curr_ctype;
+          is_first_chunk ich read_chunks curr_ctype;
+        | "EOI" ->
+          only_once ich read_chunks curr_ctype;
+          only_after ich read_chunks "SOI" curr_ctype;
+          only_after ich read_chunks "SOS" curr_ctype;
+          is_not_first_chunk ich read_chunks curr_ctype;
+
+          raise Exit
+        | "SOF0" | "SOF1"  | "SOF2"  | "SOF3"  | "SOF5"  | "SOF6"  | "SOF7"
+        | "SOF9" | "SOF10" | "SOF11" | "SOF13" | "SOF14" | "SOF15" as s ->
+          only_once ich read_chunks curr_ctype;
+          only_after ich read_chunks "SOI" curr_ctype;
+          only_before ich read_chunks "SOS" curr_ctype;
+
+          (* Frames other than SOF0 have patent issues, so are (currently) unimplemented *)
+          if s = "SOF0" then
+            let sof0 = parse_sof ich in
+            sof := Some sof0; (
```

Comment:
I find the indentation here confusing.
```ocaml
if s = "SOF0" then
  let sof0 = parse_sof ich in
  sof := Some sof0 ;

(
  (*stuff goes here*)
```

would be easier to read (the `;` shorts the then-part, so what you have above is equivalent to):
```ocaml
if s = "SOF0" then begin
  let sof0 = parse_sof ich in
  sof := Some sof0
end else begin () end ;
(
```
Not sure if that was the intention?

Also, `eprintf` below please :)